### PR TITLE
Allow primary metastore to have prefix

### LINKED ISSUE #152: Allow primary metastore to have a prefix
_As a user_
_I'd like to be able to set `primary_metastore.database-prefix` in the configuration without it being replaced by empty string_
_So that I can apply consistent prefixes across multiple waggle dance instances having different primary and federated metastores and thus re-use the same queries regardless of which waggle dance instance I am connected to._

### PR DESCRIPTION
**Pull request by AnanaMJ**
Fixes #152

### CODE REVIEW

File: waggle-dance-api/src/main/java/com/hotels/bdp/waggledance/api/model/FederatedMetaStore.java [hunk 1]
```diff
@@ -18,10 +18,9 @@
 import java.util.Collections;
 import java.util.List;
 
+import javax.validation.constraints.NotBlank;
 import javax.validation.constraints.NotNull;
 
-import org.hibernate.validator.constraints.NotBlank;
```
Comment (AnanaMJ) [inline waggle-dance-api/src/main/java/com/hotels/bdp/waggledance/api/model/FederatedMetaStore.java]: Hibernate's `@NotBlank` is deprecated. This is what the docs say:
> @deprecated use the standard {@link javax.validation.constraints.NotBlank} constraint instead

File: waggle-dance-api/src/main/java/com/hotels/bdp/waggledance/api/model/PrimaryMetaStore.java [hunk 2]
```diff
@@ -53,17 +45,18 @@ public FederationType getFederationType() {
...
 
-  @Size(min = 0, max = 0)
   @NotNull
   @Override
   public String getDatabasePrefix() {
-    // primary is always empty
-    return EMPTY_PREFIX;
+    String prefix = super.getDatabasePrefix();
+    if (prefix == null) {
+      prefix = "";
```
Comment (patduin) [inline waggle-dance-api/src/main/java/com/hotels/bdp/waggledance/api/model/PrimaryMetaStore.java]: maybe we keep using `EMPTY_PREFIX` just so it is clear that it is a valid value

**Comment by patduin** [comment]
Could you add an WaggleDanceIntegrationTest test for this?

File: CHANGELOG.md [hunk 3]
```diff
@@ -1,3 +1,7 @@
+## TBD
+### Changed
+* Allow primary metastore to have a prefix. See [#152](https://github.com/HotelsDotCom/waggle-dance/issues/152).
```
Comment (patduin) [inline CHANGELOG.md]: README also needs an update probably.

Comment (AnanaMJ) [reply CHANGELOG.md]: Done :)

File: waggle-dance-api/src/main/java/com/hotels/bdp/waggledance/api/model/PrimaryMetaStore.java (hunk 2 above)
Comment (AnanaMJ) [reply waggle-dance-api/src/main/java/com/hotels/bdp/waggledance/api/model/PrimaryMetaStore.java]: Done :)

File: waggle-dance-integration-tests/src/main/java/com/hotels/bdp/waggledance/WaggleDanceRunner.java [hunk 4]
```diff
@@ -174,6 +174,19 @@ public Builder primary(
...
 
+    public Builder primaryWithPrefix(
+        String name,
+        String remoteMetaStoreUris,
+        AccessControlType accessControlType,
+        String... writableDatabaseWhiteList) {
+      checkArgument(isNotEmpty(name));
+      checkArgument(isNotEmpty(remoteMetaStoreUris));
+      primaryMetaStore = new PrimaryMetaStore(name, remoteMetaStoreUris, accessControlType, writableDatabaseWhiteList);
+      primaryMetaStore.setLatency(8000L);
```
Comment (patduin) [inline waggle-dance-integration-tests/src/main/java/com/hotels/bdp/waggledance/WaggleDanceRunner.java]: line 182-185 can maybe be replaced with a call to `primary(name, remoteMetaStoreUris......)`

Comment (AnanaMJ) [reply waggle-dance-integration-tests/src/main/java/com/hotels/bdp/waggledance/WaggleDanceRunner.java]: Ah, thank you! That's a great suggestion :)
I've just added it

**Comment by AnanaMJ** [comment]
I’ve updated the PR with a change for `get_all_functions` in `FederatedHMSHandler` and `FederatedHMSHandlerTest`. Do you mind taking a look at it in case there's anything that I've missed?

File: waggle-dance-core/src/test/java/com/hotels/bdp/waggledance/server/FederatedHMSHandlerTest.java [hunk 5]
```diff
@@ -767,12 +768,42 @@ public void flushCache() throws TException {
 
   @Test
   public void get_all_functions() throws MetaException, TException {
```
Comment (patduin) [inline waggle-dance-core/src/test/java/com/hotels/bdp/waggledance/server/FederatedHMSHandlerTest.java]: remove this test it is not a good test anymore.

File: waggle-dance-core/src/test/java/com/hotels/bdp/waggledance/server/FederatedHMSHandlerTest.java [hunk 6]
```diff
@@ -767,12 +768,42 @@ public void flushCache() throws TException {
...
+  @Test
+  public void null_get_all_functions() throws TException {
     GetAllFunctionsResponse response = new GetAllFunctionsResponse();
     when(primaryClient.get_all_functions()).thenReturn(response);
     GetAllFunctionsResponse result = handler.get_all_functions();
     assertThat(result, is(response));
   }
 
+  @Test
+  public void prefixedPrimary_get_all_functions() throws TException {
```
Comment (patduin) [inline waggle-dance-core/src/test/java/com/hotels/bdp/waggledance/server/FederatedHMSHandlerTest.java]: rename to `get_all_functions`

Comment (AnanaMJ) [reply waggle-dance-core/src/test/java/com/hotels/bdp/waggledance/server/FederatedHMSHandlerTest.java]: Ok, done :D

File: waggle-dance-core/src/test/java/com/hotels/bdp/waggledance/server/FederatedHMSHandlerTest.java (hunk 5 above)
Comment (AnanaMJ) [reply waggle-dance-core/src/test/java/com/hotels/bdp/waggledance/server/FederatedHMSHandlerTest.java]: done

**Comment by patduin** [comment]
Could you make sure to test this with a view. Where the view references db's in the primary metastore and they get resolved correctly?

File: README.md [hunk 7]
```diff
@@ -141,7 +141,7 @@ The table below describes all the available configuration values for Waggle Danc
 | `primary-meta-store`                                    | No       | Primary MetaStore config. Can be empty but it is advised to configure it. |
 | `primary-meta-store.remote-meta-store-uris`             | Yes      | Thrift URIs of the federated read-only metastore. |
 | `primary-meta-store.name`                               | Yes      | Database name that uniquely identifies this metastore. Used internally. Cannot be empty. |
-| `primary-meta-store.database-prefix`                    | No       | This will be ignored for the primary metastore and an empty string will always be used instead. |
+| `primary-meta-store.database-prefix`                    | No       | Prefix used to access this particular metastore and differentiate databases in it from databases in another metastore. The default prefix (i.e. if this value isn't explicitly set) is empty string.|
```
Comment (massdosage) [inline README.md]: this particular -> the primary

File: README.md [hunk 8]
```diff
@@ -356,7 +356,7 @@ database is encountered that is not prefixed then the primary metastore is used
       remote-meta-store-uris: thrift://primaryLocalMetastore:9083
     federated-meta-stores:
       - name: federated
-        prefix: waggle_prod_
+        database-prefix: waggle_prod_
```
Comment (massdosage) [inline README.md]: Oh dear, was this wrong in our documentation all along?

**Comment by AnanaMJ** [comment]
I tested this with a few views, and if they don't use functions, they work as expected. If they _do_ use functions, an error like `FAILED: SemanticException [Error 10011]: Invalid function <database>.<function>` comes up. Just like the one in the issue you created (https://github.com/HotelsDotCom/waggle-dance/issues/165)

File: README.md (hunk 8 above)
Comment (patduin) [reply README.md]: oops

Comment (AnanaMJ) [reply README.md]: 😅

File: README.md (hunk 7 above)
Comment (AnanaMJ) [reply README.md]: Done :)

**Pull request merged**